libhb/decavcodec.c

```diff
@@ -835,48 +862,91 @@ reordered_hash_add(hb_work_private_t * pv, reordered_data_t * reordered)
+    // of the buffer to all.
+    hb_subtitle_t *subtitle;
+    int ii = 0, n = hb_list_count(pv->list_subtitle);
+    while (--n > 0)
     {
-        memcpy( dst, src, dstride * h );
-        return dst + dstride * h;
+        // make a copy of the buf then forward it to the decoder
+        hb_buffer_t *cpy = hb_buffer_dup(buf);
+
+        subtitle = hb_list_item(pv->list_subtitle, ii++);
+        hb_fifo_push(subtitle->fifo_in, cpy);
     }
-    int lbytes = dstride <= sstride? dstride : sstride;
-    while ( --h >= 0 )
+    subtitle = hb_list_item(pv->list_subtitle, ii);
+    hb_fifo_push( subtitle->fifo_in, buf );
+}
+
+static hb_buffer_t * cc_fill_buffer(hb_work_private_t *pv, uint8_t *cc, int size)
+{
+    int cc_count[4] = {0,};
+    int ii;
+    hb_buffer_t *buf = NULL;
+
+    for (ii = 0; ii < size; ii += 3)
+    {
+        if ((cc[ii] & 0x04) == 0)    // not valid
+            continue;
+        if ((cc[ii+1] & 0x7f) == 0 && (cc[ii+2] & 0x7f) == 0) // stuffing
+            continue;
+        int type = cc[ii] & 0x03;
+        cc_count[type]++;
+    }
+
+    // Only handles CC1 for now.
+    if (cc_count[0] > 0)
     {
-        memcpy( dst, src, lbytes );
-        src += sstride;
-        dst += dstride;
+        buf = hb_buffer_init(cc_count[0] * 2);
+        int jj = 0;
+        for (ii = 0; ii < size; ii += 3)
+        {
+            if ((cc[ii] & 0x04) == 0)    // not valid
+                continue;
+            if ((cc[ii+1] & 0x7f) == 0 && (cc[ii+2] & 0x7f) == 0) // stuffing
+                continue;
+            int type = cc[ii] & 0x03;
+            if (type == 0)
+            {
+                buf->data[jj++] = cc[ii+1];
+                buf->data[jj++] = cc[ii+2];
+            }
+        }
     }
-    return dst;
+    return buf;
 }
 
 // copy one video frame into an HB buf. If the frame isn't in our color space
 // or at least one of its dimensions is odd, use sws_scale to convert/rescale it.
 // Otherwise just copy the bits.
 static hb_buffer_t *copy_frame( hb_work_private_t *pv )
 {
-    AVCodecContext *context = pv->context;
-    int w, h;
-    if ( ! pv->job )
+    reordered_data_t * reordered = NULL;
+    hb_buffer_t      * out;
+
+#ifdef USE_QSV
+    // no need to copy the frame data when decoding with QSV to opaque memory
+    if (pv->qsv.decode &&
+        pv->qsv.config.io_pattern == MFX_IOPATTERN_OUT_OPAQUE_MEMORY)
     {
-        // HandBrake's video pipeline uses yuv420 color.  This means all
-        // dimensions must be even.  So we must adjust the dimensions
-        // of incoming video if not even.
-        w = context->width & ~1;
-        h = context->height & ~1;
+        out = hb_frame_buffer_init(frame->format, frame->width, frame->height);
+        hb_avframe_set_video_buffer_flags(buf, frame, (AVRational){1,1});
+
+        out->qsv_details.qsv_atom = pv->frame->data[2];
+        out->qsv_details.ctx      = pv->job->qsv.ctx;
     }
     else
+#endif
     {
```

Comment:
When USE_QSV is not defined, these braces will still be present, fwiw.

Reply:
Yes, the braces are reqired when USE_QSV is defined.  And do no harm when not.  The alternative is to repeat the code formatted without the braces which I dislike because then you have 2 copies of the same code that you have to keep in sync.

Reply:
Yep, just wanted to make sure it wasn't an oversight. There's no beautiful way in C.